merlin/models/tf/experimental/sample_weight.py

```diff
@@ -0,0 +1,165 @@
+        candidate_tag_id=Tags.ITEM_ID,
+        **kwargs,
+    ):
+        """Generate the sample weights for positive and sampled
+        negative candidates.
+        This class can be used as a `post` of the `ContrastiveOutput`
+        instance.
+
+        Parameters
+        ----------
+        pos_class_weight: Union[str, tf.Tensor, float]
+            The positive candidates weights. This class supports three types of weights:
+            * str: the column name of the input feature to use as the sample weight.
+            In this case, the weights are dynamic and change based on the input
+            (user-candidate) interaction.
+            * tf.Tensor: a weight vector of shape (#unique-candidates,). In this case,
+            the weights are specific to each candidate and are the same for all
+            users' interactions with the given candidate.
+            * float: use the same weight value for all positives candidates. In this case,
+            the weights are static for all (user-candidate) interactions.
+        neg_class_weight:  Union[tf.Tensor, float]
+            The negative candidates weights. The class supports two types of weights:
+            * tf.Tensor: a weight vector of shape (#unique-candidates,). In this case,
+            the weights are specific to each sampled negative candidate.
+            * float: Use the same weight value for all negative candidates.
+        schema: Schema, optional
+            The `Schema` with candidate id feature,
+            by default None
+        candidate_tag_id: Tag, optional
+            The tag of the candidate id column. This feature is needed to look-up for the
+            positive candidates weights when `pos_class_weight` is a `tf.Tensor` object.
+            by default 'item_id'
+
+        Example usage::
+            outputs = mm.ContrastiveOutput(
+                        DotProduct(),
+                        schema=data.schema.select_by_tag(Tags.ITEM_ID),
+                        negative_samplers="in-batch",
+                        store_negative_ids=True,
+                        post=ContrastiveSampleWeight(
+                            pos_class_weight='interaction-weight',
+                            neg_class_weight=0.5,
+                        ),
+                    )
+        """
+        super().__init__(**kwargs)
+        if schema:
+            self.set_schema(schema)
+        self.pos_class_weight = pos_class_weight
+        self.neg_class_weight = neg_class_weight
+        self.candidate_tag_id = candidate_tag_id
+        self.candidate_id_name = self.schema.select_by_tag(self.candidate_tag_id).first.name
+
+    def call(
+        self,
+        outputs: Prediction,
+        features,
+        training=False,
+        testing=False,
+    ) -> Prediction:
+        """Update the predictions returned by the
+        ConstrastiveOutput with a 2-D sample weights
+        for negative and positive candidates.
+        """
+        if training:
```

Comment:
Why do you apply `sample_weight` only during `training`?
During evaluation (`testing == True`) I think we should also apply the sample weights, so that `weighted_metrics` (if provided to `model.compile()`) take the sample weights into account.

Reply:
I didn’t know about the weighted_metrics option, as usually the weights are not available at inference and the evaluation metrics should also account for this scenario. But I can see that, in some situations, applying the weights during evaluation can be useful (to take into account the importance of getting a specific set of item-ids right). Thank you for pointing out this option, I included the evaluation mode too (testing==True)